x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/ResourceUtils.java

```diff
@@ -25,16 +28,19 @@ static byte[] loadVersionedResourceUTF8(
         Map<String, String> variables
     ) {
         try {
-            String content = loadResource(clazz, name);
+            String content = loadResourceWithFallback(clazz, name);
             content = TemplateUtils.replaceVariables(content, String.valueOf(version), versionProperty, variables);
             return content.getBytes(StandardCharsets.UTF_8);
         } catch (IOException e) {
             throw new RuntimeException(e);
         }
     }
 
-    public static String loadResource(Class<?> clazz, String name) throws IOException {
+    static String loadResourceWithFallback(Class<?> clazz, String name) throws IOException {
         InputStream is = clazz.getResourceAsStream(name);
+        if (is == null) {
+            is = TemplateResources.class.getResourceAsStream("/"+clazz.getPackageName()+name);
```

Comment:
[**For reviewers**] I decided to use the package name as the folder in the template-resources so that we don't have to bloat the interface with fallback root path. Let me know if there is a better suggestion to do this.

Reply:
Just to make sure I understand: this would mean that we'd have something like this on disk?

`/path/to/override/template-resources/org.elasticsearch.xpack.apmdata/component-templates/whatever.yaml`

Is that right? Do plugins live in their own Java module? If so would it make sense to use the module name, rather than the package name?

Is it possible to add a YAML REST test to apm-data showing the overrides working?

Reply:
> Should it be the other way around? i.e. look for an override in template-resources first, and in the plugin second?

Hmm, this does make sense w.r.t. what the template resource is meant to do. I will make the changes.

> Is that right? 

Yes, I have a sample for this in https://github.com/elastic/elasticsearch/pull/112614/files

> Do plugins live in their own Java module?

Not sure if I understand "Java module" correctly and my Java9++ is not that good but from what I can see, we haven't modularized apm-data plugin. I think it should be as simple as adding the correct module-info. If we do that, then we can use the module name. I will give it a shot.

> Is it possible to add a YAML REST test to apm-data showing the overrides working?

We are not overriding in this PR, the fallback files are required by the plugin, and failing to locate them would be fatal for starting ES -- so, they are already be covered by the tests. Are you thinking about adding explicit tests to check if all the files are installed?

Reply:
Maybe I'm totally misunderstanding the approach. What I thought we were going to do is:

1. Add the "-fallback@lifecycle" component templates in the plugin, in a way that would work with "stateful" Elasticsearch OOTB
2. Allow these to be overridden by placing a component template of the same name in template-resources; if that exists, prefer using that

(I was not expecting to see any templates added to template-resources in https://github.com/elastic/elasticsearch/pull/112614)

In terms of YAML REST tests, I was thinking we would:
 1. Add test cases for the scenarios laid out in https://github.com/elastic/apm-server/issues/13898#issuecomment-2326277518
 2. Add another test case that adds an override

... but thinking about it some more, I'm not sure if (2) is possible. I had in mind that the overrides were on disk, but I think they're also embedded in a jar? Just it's a different jar in serverless vs. stateful.

Reply:
> Allow these to be overridden by placing a component template of the same name in template-resources; if that exists, prefer using that

Ah, I see the misunderstanding. The overrides for serverless are performed with files in https://github.com/elastic/elasticsearch-serverless/tree/main/libs/serverless-xpack-template-resources. I am not sure if we can override for serverless using template-resources (not sure how we can trigger the if/else condition i.e. if we are on serverless load this...). The approach suggested to us is to provide the fallback to ILM with the default apm-data plugin and override the fallback files for serverless to be empty.

> ... but thinking about it some more, I'm not sure if (2) is possible. I had in mind that the overrides were on disk, but I think they're also embedded in a jar? Just it's a different jar in serverless vs. stateful.

Exactly. Also, we could have added test if some version of apm-data supported ILM (by upgrading from that version) but with the current state of things, it is not straightforward and will require creating datastreams with ILMs and then applying the apm-data plugin... BTW, the current PR is only about introducing the fallback mechanism for loading template resources in `YamlTemplateRegistry`, we can discuss more on the tests in the PR where we introduce the ILM files.

Reply:
I think this looks good. I think using the module name should be a bit less brittle than the package name.